solr/solrj/src/java/org/apache/solr/client/solrj/impl/CloudSolrClient.java

```diff
@@ -117,7 +119,12 @@ public abstract class CloudSolrClient extends SolrClient {
           // UpdateParams.ROLLBACK
           );
 
-  protected volatile Object[] locks = objectList(3);
+  private final ConcurrentHashMap<String, CompletableFuture<DocCollection>> collectionRefreshes =
```

Comment:
Instead of creating an additional cache, I think it would be better to integrate into StateCache, which is already collection-keyed, already has eviction of old stuff.